### PR TITLE
Fix linking issues with HDF5 when using DLAF as dependency and enable HDF5 in CI

### LINKED ISSUE #981: Linking issues with standalone miniapps and HDF5
When building DLA-Future with `+hdf5` and subsequently building the miniapps as a standalone project
```
cmake .. -DCMAKE_PREFIX_PATH=`spack -e <ENV> location -i dla-future`
```
(as done in CI), linking errors related to HDF5 (`-lhdf5_cpp-shared`) appear.

The same issue appears when trying to build CP2K with `dla-future +hdf5`.

### PR DESCRIPTION
**Pull request by RMeli**
This PR adds a missing variable to `DLAFConfig.cmake.in`, whose absence resulted in linking issues when using `dla-future +hdf5` as a dependency (CP2K, building the miniapps as standalone, ...). Fix #981.

This issue could have been [spotted by CI](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/5068361053), so I'm enabling the `+hdf5` variant in CI as well.

### CODE REVIEW

**Comment by RMeli** [comment]
cscs-ci run

**Review by msimberg — APPROVED** [review]

**Comment by msimberg** [comment]
This may be a problem though: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/5070078215#L370. @albestro do the unit tests have special requirements?

**Comment by RMeli** [comment]
Yep, I've seen the CI failure. (I change this PR from draft to ready when I saw that the building phase was all green, and did not wait for all the other tests to finish, sorry about that).

I think the only requirement is that `test_matrix_hdf5.h5` can be created/opened with `FileHDF5::FileMode::readwrite` access.

**Comment by albestro** [comment]
> This may be a problem though: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/5070078215#L370. @albestro do the unit tests have special requirements?

IIRC what @RMeli said is correct and it should be the only requirement.
 
> I think the only requirement is that `test_matrix_hdf5.h5` can be created/opened with `FileHDF5::FileMode::readwrite` access.

**Review by albestro — COMMENTED** [review]
Thanks for the fix!

About enabling HDF5 in CI: I would evaluate to enable it just in a (small) set of configurations. My main reason (not very strong) is that the main use case for a user is WITHOUT HDF5.

Probably an option might be to enable hdf5 just in a debug and a release configurations? or just a debug one?

@rasolca @msimberg @RMeli what do you think?

**Comment by RMeli** [comment]
> Probably an option might be to enable hdf5 just in a debug and a release configurations? or just a debug one?

As discussed on Monday, I removed the `+hdf5` variant from the release builds so that we test DLA-Future without this variant too. 

Following the same discussion, I also moved the `hdf5` variant amongst the ones we don't want to propagate upstream, so that we remember to remove it before updating the Spack package.

**Comment by RMeli** [comment]
cscs-ci run

File: test/unit/matrix/test_matrix_hdf5.cpp [hunk 1]
```diff
@@ -33,7 +33,7 @@ using dlaf::matrix::internal::FileHDF5;
 template <typename T>
 class MatrixHDF5Test : public ::testing::Test {
 protected:
-  MatrixHDF5Test() : world(MPI_COMM_WORLD), filepath("test_matrix_hdf5.h5") {
+  MatrixHDF5Test() : world(MPI_COMM_WORLD), filepath("/tmp/test_matrix_hdf5.h5") {
```
Comment (msimberg) [inline test/unit/matrix/test_matrix_hdf5.cpp:36]: Did you find out if there were restrictions on the directory where this test was run inside the CI job?

**Review by msimberg — APPROVED** [review]

File: test/unit/matrix/test_matrix_hdf5.cpp (hunk 1 above)
Comment (RMeli) [reply test/unit/matrix/test_matrix_hdf5.cpp:36]: Still experimenting. But running `ctest` directly on `daint` from `$HOME/git/DLA-Future/build` results in the same issue.

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

File: test/unit/matrix/test_matrix_hdf5.cpp [hunk 2]
```diff
@@ -30,10 +31,21 @@
 using namespace dlaf;
 using dlaf::matrix::internal::FileHDF5;
 
+constexpr auto env_hdf5_output_path = "DLAF_HDF5_TEST_OUTPUT_PATH";
+const std::filesystem::path filename = "test_matrix_hdf5.h5";
+
+// Get path from environment variable if defned, otherwise return empty path
```
Comment (RMeli) [inline test/unit/matrix/test_matrix_hdf5.cpp:37]:
```suggestion
// Get path from environment variable if defined, otherwise return empty path
```

**Comment by RMeli** [comment]
HDF5 test finally working in CI: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/5183605357#L241

Thanks @finkandreas for the help!

**Review by msimberg — APPROVED** [review]

**Review by rasolca — APPROVED** [review]

**Comment by rasolca** [comment]
cscs-ci run

**Pull request merged**